internal/scheduler/scheduler.go

```diff
@@ -484,19 +526,69 @@ func (s *Scheduler) generateUpdateMessagesFromJob(job *jobdb.Job, jobRunErrors m
 				},
 			}
 			events = append(events, jobSucceeded)
-		} else if lastRun.Failed() {
-			job = job.WithFailed(true).WithQueued(false)
-			runError := jobRunErrors[lastRun.Id()]
-			jobErrors := &armadaevents.EventSequence_Event{
-				Created: s.now(),
-				Event: &armadaevents.EventSequence_Event_JobErrors{
-					JobErrors: &armadaevents.JobErrors{
-						JobId:  jobId,
-						Errors: []*armadaevents.Error{runError},
+		} else if lastRun.Failed() && !job.Queued() {
+			requeueJob := lastRun.Returned() && job.NumAttempts() < s.maxAttemptedRuns
+
+			if requeueJob && lastRun.RunAttempted() {
+				jobWithAntiAffinity, schedulable, err := s.addNodeAntiAffinitiesForAttemptedRunsIfSchedulable(job)
+				if err != nil {
+					return nil, errors.Errorf("unable to set node anti-affinity for job %s because %s", job.GetId(), err)
+				} else {
+					if schedulable {
+						job = jobWithAntiAffinity
+					} else {
+						// If job is not schedulable with anti-affinity added. Do not requeue it and let it fail
+						requeueJob = false
+					}
+				}
+			}
+
+			if requeueJob {
+				job = job.WithUpdatedRun(lastRun.WithFailed(false))
+				job = job.WithQueued(true)
+				job = job.WithQueuedVersion(job.QueuedVersion() + 1)
+
+				requeueJobEvent := &armadaevents.EventSequence_Event{
+					Created: s.now(),
+					Event: &armadaevents.EventSequence_Event_JobRequeued{
+						JobRequeued: &armadaevents.JobRequeued{
+							JobId:                jobId,
+							SchedulingInfo:       job.JobSchedulingInfo(),
+							UpdateSequenceNumber: job.QueuedVersion(),
+						},
 					},
-				},
+				}
+
+				events = append(events, requeueJobEvent)
+			} else {
+				runError := jobRunErrors[lastRun.Id()]
+				job = job.WithFailed(true).WithQueued(false)
+				if lastRun.Returned() {
+					errorMessage := fmt.Sprintf("Maximum number of attempts (%d) reached - this job will no longer be retried", s.maxAttemptedRuns)
+					if job.NumAttempts() < s.maxAttemptedRuns {
+						errorMessage = fmt.Sprintf("Job was attempeted %d times, and has been tried once on all nodes it can run on - this job will no longer be retried", job.NumAttempts())
```

Comment:
```suggestion
						errorMessage = errors.Sprintf("Job was attempeted %d times, and has been tried once on all nodes it can run on - this job will no longer be retried", job.NumAttempts())
```